### PR TITLE
Refactoring `server.py` and migrating to `python-osc`

### PR DESCRIPTION
**Pull request by Itaborala**
This should solve #9 and #11 

Changes:
- The Server class was refactored to use `python-osc` instead of `pyliblo` for OSC
- Handler structure was simplified
- The code should no longer require `liblo`, `pyliblo` and `Cython` .
- Version on `setup.py` was changed to 0.0.6, preparing for publishing

For approving this pull request, we need to make sure that the new server is fully backwards compatible with all other classes and methods and no bugs (originated by the refactoring) are present.

### CODE REVIEW

File: README.md
Comment (ideoforms) [inline README.md]: No need to keep the old docs - liblo is now deprecated, so feel free to replace the instructions (and no need to mention liblo at all).

File: supercollider/buffer.py
Comment (ideoforms) [inline supercollider/buffer.py]: Can delete these commented lines

File: supercollider/old_server.py
Comment (ideoforms) [inline supercollider/old_server.py]: Can delete this obsolete file

File: supercollider/server.py
Comment (ideoforms) [inline supercollider/server.py]: Remove commented line

File: supercollider/server.py
Comment (ideoforms) [inline supercollider/server.py]: Remove commented line

File: supercollider/server.py
Comment (ideoforms) [inline supercollider/server.py]: `num`: Should be a more self-explanatory argument name - what does `num` represent?

File: supercollider/server.py
Comment (ideoforms) [inline supercollider/server.py]: I'm not convinced we need to keep this TODO - I think this method is fine! :-)

File: supercollider/server.py
Comment (ideoforms) [inline supercollider/server.py]: Bad practice to have magic numbers in the code (the original used `globals.RESPONSE_TIMEOUT` - could this still be used?)

**Review by ideoforms — CHANGES_REQUESTED** [review]
Just spotted this PR - thanks so much!
I haven't yet tested or run the code, so this first review is mostly cosmetics. I'll follow up with a second review that takes a closer look at performance, tests etc.

**Comment by spyridonpallis** [comment]
Thanks for updating this guys. Any timeline on when this will be completed?

**Comment by Itaborala** [comment]
Hi @spyridonpallis , I've been very busy this month working on a time sensitive project. I should get back to this next week and hopefully we can have it completed in May. :)

File: README.md
Comment (Itaborala) [reply README.md]: I would still suggest to keep some information about this change in the README file. I made a second version: 861e77fe1a29c27d475489072747d3bb678bca19 using fewer lines. What do you think?

File: supercollider/server.py
Comment (Itaborala) [reply supercollider/server.py]: Absolutely. I removed it during the refactoring process and forgot to put it back

**Comment by Itaborala** [comment]
@ideoforms Just letting you know I'm back to this, fully willing to finishing and having this publishes this month, if possible

**Comment by ideoforms** [comment]
@Itaborala That would be great!
I have just cloned from your fork. Unit tests are passing except for `test_server_get_version`, which looks pretty simple to fix:

```
    def get_version(self):
        self._send_msg("/version")

>       return self._await_response("/version.reply", None, version_handler)
E       NameError: name 'version_handler' is not defined

supercollider/server.py:114: NameError
```

Let me know when you're at a point when this is ready for review with passing tests, and I'll take a final look over.

**Comment by ideoforms** [comment]
@Itaborala Let me know if you're still intending on finishing this up. No worries if not; I'd be happy to merge the changes and then tidy themself up myself. Thanks!

**Comment by Itaborala** [comment]
Hi @ideoforms , sorry, those busy days became some busy months..
I can help you finish this up right now. I fixed the error, now all tests should work.

Are there any other changes/additions to be made?

File: supercollider/server.py
Comment (ideoforms) [inline supercollider/server.py]: Commented-out code, should be deleted

File: supercollider/server.py
Comment (ideoforms) [inline supercollider/server.py]: I'm not keen on having instance methods for these "handler" functions, for a few reasons:
 - it clutters up the `Server` namespace with methods that are only designed to be used internally (try `pydoc supercollider.Server` and you will see all of these methods showing up)
 - it distributes the message-parsing logic over different areas of the code; e.g. in the case of `get_version`, all of the logic around dict construction was previously encapsulated in the `get_version` function
 - in the case of `buf_handler`, it means that other classes (e.g. `Buffer`) also have to utilise message parsing functions from Server

I'd prefer to return to the previous approach of implementing local `_handler` functions within the methods that utilise them. Thanks!

File: supercollider/server.py
Comment (ideoforms) [inline supercollider/server.py:107]: Please restore the pydocs for `get_version` and `get_status` - they seem to have been lost in this fork

File: supercollider/test.scd
Comment (ideoforms) [inline supercollider/test.scd:1]: Looks like a scratch file, should be deleted

**Review by ideoforms — CHANGES_REQUESTED** [review]
Thanks for the PR! Just ran the tests and they're all passing fine. 🎉 

I've added a few comments, main one being about restoring the previous approach to handling callback arguments - then good to merge.

File: supercollider/test.scd
Comment (Itaborala) [reply supercollider/test.scd:1]: Solved in [2c5faf0](https://github.com/ideoforms/python-supercollider/pull/12/commits/2c5faf0bbae04e294ec530f3bfe7c3533edd629f)

File: supercollider/server.py
Comment (Itaborala) [reply supercollider/server.py:107]: Solved in [9bd584f](https://github.com/ideoforms/python-supercollider/pull/12/commits/9bd584f53a5ec7a0eed072ca67869c326910dcdc)

File: supercollider/server.py
Comment (Itaborala) [reply supercollider/server.py]: Solved in [9bd584f](https://github.com/ideoforms/python-supercollider/pull/12/commits/9bd584f53a5ec7a0eed072ca67869c326910dcdc)

File: supercollider/server.py
Comment (Itaborala) [reply supercollider/server.py]: Done. I have made all handlers internal again in [0be7cd2](https://github.com/ideoforms/python-supercollider/pull/12/commits/0be7cd2a75c6e378e7968ac64c7aca7b7bfbf96b)

I've also noticed that the cases where blocking=False were still expecting `pyliblo` handlers. [910e857](https://github.com/ideoforms/python-supercollider/pull/12/commits/910e8578e8da3afefdbf9b12a30d440eedc5afab)

File: README.md
Comment (Itaborala) [reply README.md]: Solved in [56e72bb](https://github.com/ideoforms/python-supercollider/pull/12/commits/56e72bb9b7ec30ab1c07e69eb9605b02ef0d49fb)

**Comment by Itaborala** [comment]
@ideoforms I've addressed all requests. Tests are working for me too.

Additionally, I have noticed that in `buffer.py` and `synth.py` there were `if blocking` statements which were still using pyliblo handlers, although the condition `blocking=False` is never tested. I don't know of any usecases for `blocking=False`, but I have now addressed it.

To account for `python-osc`, I have changed `server.add_handler()` into `server.dispatcher.map()`, and included a method in `server.py` that can clear, or 'flush' all non_blocking handlers (remember that the `_await_response` method unmaps the handles after the timeout).
See [910e857](https://github.com/ideoforms/python-supercollider/pull/12/commits/910e8578e8da3afefdbf9b12a30d440eedc5afab) for this addition

**Review by ideoforms — APPROVED** [review]
Looks great @Itaborala! Merging now, thanks for all your work on this.

**Comment by ideoforms** [comment]
@Itaborala I have just added tests for the `blocking=False` situations. These are needed if you have a large number of simultaneous queries happening and don't want to hold up the foreground thread whilst awaiting responses.

On that note, I realised that part of the query system will now be broken following the move to python-osc. The callback system previously had callbacks indexed not only by the OSC address (`/n_set`) but also by the synth ID and the parameter being queried. This meant that several getter queries could be simultaneously deployed, and their callbacks would all be triggered when the responses were received. Now, callbacks are indexed only by the OSC address - meaning that each new `/n_set` callback registered will overwrite the previous ones. I realised there wasn't a test to catch this case which is how it slipped through, but I spotted it because `match_args` is now unused in `_await_response()`.

I don't think it should hold up a new release as it's probably a feature that's only needed in a few cases, but I'll add an issue for it.

**Comment by ideoforms** [comment]
I've now published this to PyPi as release v0.0.6:
https://pypi.org/project/supercollider/

**Pull request merged**
